File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        def check(p, q):
            if not p and not q:
                return True
            
            if (p and not q) or (not p and q):
                return False
            
            if p and q and p.val != q.val:
                return False
            
            return check(p.left, q.left) and check(p.right, q.right)
        
        def traverse(root):
            if not root:
                return False
 
            if check(root, subRoot):
                return True
            
            return traverse(root.left) or traverse(root.right) 

        return traverse(root)

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-3.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func buildTree(preorder []int, inorder []int) *TreeNode {
    hashMap := map[int]int{}
    idx := 0

    for i:= 0; i < len(inorder); i++ {
        hashMap[inorder[i]] = i
    }

    var construct func(l int, r int) *TreeNode 
    construct = func(l int, r int) *TreeNode {
        if l > r {
            return nil
        }

        root := preorder[idx]
        idx++

        mid := hashMap[root]
        node := &TreeNode{ Val: root}

        node.Left = construct(l, mid - 1)
        node.Right = construct(mid + 1, r)

        return node
    }

    return construct(0, len(preorder) - 1)
}

